File: Data Structures & Algorithms/serialize-and-deserialize-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# 1#2#3#N#N#4#5# where # represents end of digits and N reprsetns null values
from collections import deque

class Codec:
    
    # Encodes a tree to a single string.
    def serialize(self, root: Optional[TreeNode]) -> str:
            res = []

            def dfs(node) -> None:
                if not node:
                    res.append("N")
                    return 
                res.append(str(node.val))
                dfs(node.left)
                dfs(node.right)
            dfs(root)
            return "#".join(res)

    # Decodes your encoded data to tree.
    def deserialize(self, data: str) -> Optional[TreeNode]:
        splitted = data.split("#")  # 1 , 2 , 3 , N , N, 4 
        queue = deque(splitted)

        def convertToTree(queue)-> TreeNode:
            if  queue:
                val = queue.popleft()
    
                if val == 'N':
                    return None

                node = TreeNode(val)

                node.left = convertToTree(queue)
                node.right = convertToTree(queue)
            return node
        return convertToTree(queue)

File: Data Structures & Algorithms/depth-of-binary-tree/submission-0.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    maxDepth(root) {
        const q = new Queue();
        if (root !== null) {
            q.push(root);
        }

        let level = 0;
        while (q.size() > 0) {
            const size = q.size();
            for (let i = 0; i < size; i++) {
                const node = q.pop();
                if (node.left !== null) {
                    q.push(node.left);
                }
                if (node.right !== null) {
                    q.push(node.right);
                }
            }
            level++;
        }
        return level;
    }
}